arch/xtensa/include/kernel_arch_func.h

```diff
@@ -83,24 +83,61 @@ static inline void arch_cohere_stacks(struct k_thread *old_thread,
 	size_t nsz    = new_thread->stack_info.size;
 	size_t nsp    = (size_t) new_thread->switch_handle;
 
+	/* The "live" area (the region between the switch handle,
+	 * which is the stack pointer, and the top of the stack
+	 * memory) of the inbound stack needs to be invalidated: it
+	 * may contain data that was modified on another CPU since the
+	 * last time this CPU ran the thread, and our cache may be
+	 * stale.
+	 *
+	 * The corresponding "dead area" of the inbound stack can be
+	 * ignored.  We may have cached data in that region, but by
+	 * definition any unused stack memory will always be written
+	 * before being read (well, unless the code has an
+	 * uninitialized data error) so our stale cache will be
+	 * automatically overwritten as needed.
+	 */
 	z_xtensa_cache_inv((void *)nsp, (nstack + nsz) - nsp);
 
-	/* FIXME: dummy initializion threads don't have stack info set
-	 * up and explode the logic above.  Find a way to get this
-	 * test out of the hot paths!
+	/* Dummy threads appear at system initialization, but don't
+	 * have stack_info data and will never be saved.  Ignore.
 	 */
 	if (old_thread->base.thread_state & _THREAD_DUMMY) {
 		return;
 	}
 
-	/* In interrupt context, we have a valid frame already from
-	 * the interrupt entry code, but for arch_switch() that hasn't
-	 * happened yet.  It will do the flush itself, we just have to
-	 * calculate the boundary for it.
+	/* For the outbound thread, we obviousy want to flush any data
+	 * in the live area (for the benefit of whichever CPU runs
+	 * this thread next).  But we ALSO have to invalidate the dead
+	 * region of the stack.  Those lines may have DIRTY data in
+	 * our own cache, and we cannot be allowed to write them back
+	 * later on top of the stack's legitimate owner!
+	 *
+	 * This work comes in two flavors.  In interrupts, the
+	 * outgoing context has already been saved for us, so we can
+	 * do the flush right here.  In direct context switches, we
+	 * are still using the stack, so we do the invalidate of the
+	 * bottom here, (and flush the line containing SP to handle
+	 * the overlap).  The remaining flush of the live region
+	 * happens in the assembly code once the context is pushed, up
+	 * to the stack top stashed in a special register.
 	 */
 	if (old_switch_handle != NULL) {
 		z_xtensa_cache_flush((void *)osp, (ostack + osz) - osp);
+		z_xtensa_cache_inv((void *)ostack, osp - ostack);
 	} else {
+		/* When in a switch, our current stack is the outbound
+		 * stack.  Flush the single line containing the stack
+		 * bottom (which is live data) before invalidating
+		 * everything below that.  Remember that the 16 bytes
+		 * below our SP are the calling function's spill area
+		 * and may be live too.
+		 */
+		__asm__ volatile("mov %0, a1" : "=r"(osp));
+		osp -= 16;
+		z_xtensa_cache_flush((void *)osp, 1);
```

Comment:
The rest of the lines get flushed in the middle of `arch_switch()`.  This one line is here because there may be overlap in that line between the bytes that are live at the bottom of the stack and the bytes **below** the stack that have to be invalidated.  We have to be sure that there is no dirty data in the stack, but can't accidentally invalidate real data at the bottom.  So we flush the one line where there may be overlap.

Reply:
@andyross so the case of the spill area crossing 2 cache lines is impossible? Can our current stack pointer be something like 0x1008 then you'd need to flush two lines: at 0xfc0 and 0x1000 - assuming 0x40 byte cache lines?

Reply:
The caller spill area just moves the boundary.  Again, this isn't the whole cache operation, the rest is completed later, in `arch_switch()`.   The theory is that we have two regions of interest:

1. The top of the stack with live data that has to be flushed
2. The bottom of the stack with stale data that might be dirty and must be flushed or invalidated, we can't have dirty data in the cache once the thread stops running (because it will end up being flushed out later on top of the same thread running on another CPU).

This code here, in `arch_cohere_stacks()`, is only concerned with region 2 (because we're still running C code and dirtying our own stack, we can't flush yet -- that happens in assembly right before we switch stacks into the new context).

But there's a complexity, because there can be one cache line that has bytes from region 1 **and** region 2.  So we flush the top line instead of invalidating it.

Reply:
Oh sorry, and you had a question above that I skipped: the "caller" in this case refers to the ABI calling context -- the last spot where a CALLn instruction was issued, not necessarily a C function.

Reply:
> so the case of the spill area crossing 2 cache lines is impossible?

Spill area is 16 bytes in size and a cache line may be as small as 4 bytes. So in the worst case the spill area can cover 4 cache lines.

> Can our current stack pointer be something like 0x1008

According to xtensa ABI stack pointer must be aligned to 16 bytes.